### PR TITLE
chore: Convert heap allocation to optional

### PR DESCRIPTION
**Pull request by jadebenn**
swap out the petcomponent precondition expression for an optional
doing this here because the petfixes PR is already big enough

### CODE REVIEW

**Review by aronwk-aaron — APPROVED** [review]

File: dGame/dComponents/PetComponent.cpp [hunk 1]
```diff
@@ -158,7 +157,7 @@ void PetComponent::OnUse(Entity* originator) {
 		return;
 	}
 
-	if (m_Preconditions != nullptr && !m_Preconditions->Check(originator, true)) {
+	if (m_Preconditions.has_value() && !m_Preconditions->Check(originator, true)) {
```
Comment (EmosewaMC) [inline dGame/dComponents/PetComponent.cpp:160]:
```suggestion
	if (m_Preconditions && !m_Preconditions->Check(originator, true)) {
```
optional should be implicitly convertible to bool so the has_value can be dropped

File: dGame/dComponents/PetComponent.h [hunk 2]
```diff
@@ -349,7 +349,7 @@ class PetComponent final : public Component
 	/**
 	 * Preconditions that need to be met before an entity can tame this pet
 	 */
-	PreconditionExpression* m_Preconditions;
+	std::optional<PreconditionExpression> m_Preconditions = std::nullopt;
```
Comment (EmosewaMC) [inline dGame/dComponents/PetComponent.h:352]:
```suggestion
	std::optional<PreconditionExpression> m_Preconditions;
```
This should achieve the same result?

Comment (jadebenn) [reply dGame/dComponents/PetComponent.h:352]: If I'm going to default-initialize I think I should actually go with
```suggestion
	std::optional<PreconditionExpression> m_Preconditions{};
```

**Review by aronwk-aaron — APPROVED** [review]

File: dGame/dComponents/PetComponent.cpp (hunk 1 above)
Comment (EmosewaMC) [reply dGame/dComponents/PetComponent.cpp:160]: The codebase standard has been pretty consistent with utilizing operator bools on optional.  Not sure why specifically here you are drawing the line with readability.

**Review by EmosewaMC — APPROVED** [review]

**Pull request merged**